core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -1,126 +1,84 @@
 package org.infinispan.interceptors.locking;
 
 import java.util.Collection;
+import java.util.Collections;
 
 import org.infinispan.commands.CommandsFactory;
 import org.infinispan.commands.DataCommand;
 import org.infinispan.commands.FlagAffectedCommand;
 import org.infinispan.commands.control.LockControlCommand;
 import org.infinispan.commands.tx.PrepareCommand;
 import org.infinispan.commands.write.DataWriteCommand;
+import org.infinispan.context.Flag;
 import org.infinispan.context.InvocationContext;
 import org.infinispan.context.impl.FlagBitSets;
 import org.infinispan.context.impl.TxInvocationContext;
 import org.infinispan.factories.annotations.Inject;
-import org.infinispan.interceptors.InvocationSuccessFunction;
-import org.infinispan.statetransfer.StateTransferManager;
-import org.infinispan.transaction.impl.LocalTransaction;
+import org.infinispan.interceptors.distribution.TxDistributionInterceptor;
 import org.infinispan.util.logging.Log;
 import org.infinispan.util.logging.LogFactory;
 
 /**
- * Locking interceptor to be used by pessimistic caches.
- * Design note: when a lock "k" needs to be acquired (e.g. cache.put("k", "v")), if the lock owner is the local node,
- * no remote call is performed to migrate locking logic to the other (numOwners - 1) lock owners. This is a good
- * optimisation for  in-vm transactions: if the local node crashes before prepare then the replicated lock information
- * would be useless as the tx is rolled back. OTOH for remote hotrod/transactions this additional RPC makes sense because
- * there's no such thing as transaction originator node, so this might become a configuration option when HotRod tx are
- * in place.
- *
- * Implementation note: current implementation acquires locks remotely first and then locally. This is required
- * by the deadlock detection logic, but might not be optimal: acquiring locks locally first might help to fail fast the
- * in the case of keys being locked.
+ * Locking interceptor to be used by pessimistic transactions.
+ * <p>
+ * Design (from 9.1):
+ * <p>
+ * This interceptor is skipped when {@link Flag#SKIP_LOCKING} is used or when all affected keys are already locked.
+ * Read-only commands can acquire the locks when {@link Flag#FORCE_WRITE_LOCK}.
+ * <p>
+ * The new design has 2 concepts. The primary lock and the backup lock. The primary lock is acquired in the primary
+ * owner and only on transaction can acquire at each give time. The backup lock exists in the backup owner and multiple
+ * transactions can acquire simultaneously. The backup lock is associated to the current topology id and the primary
+ * lock can only be acquired if doesn't exists any backup lock with older topology id.
+ * <p>
+ * Since the primary owner doesn't acquire backup locks, they are only used when the primary ownership of a key is
+ * changed.
+ * <p>
+ * The new algorithm will always acquire both. So, lock acquisition (single or multiple keys) will generate a RPC to all
```

Comment:
This is unclear: acquire both locks on each node, or just primary on primaries and backup on the others? Or does it just mean that a LockCommand will contact all owners? Don't try to describe "new" algorithm assuming that the "old" i known, just describe how it is.

Reply:
This new algorithm sounds very much like the old algorithm, the only difference seems to be the removal of `isSingleKeyAndLocal` optimization. (Which I believe was supposed to mean a single key in the transaction, so it was broken all along.)

@pruivo are you sure we can't fix that optimization, maybe at the expense of some more work during state transfer? Requiring an extra RPC when the local node is the primary owner would definitely impact the latency of HotRod writes to pessimistic caches.